File: 0912-sort-an-array/0912-sort-an-array.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        heapify(nums)
        return [heappop(nums) for i in range(len(nums))]


# class Solution:
#     def sortArray(self, nums: List[int]) -> List[int]:
        
#         def merge(a, b):
#             n, m = len(a), len(b)
#             ans = []
            
#             i = j = 0
            
#             while i<n and j<m:
#                 if a[i]<=b[j]:
#                     ans.append(a[i])
#                     i+=1
#                 else:
#                     ans.append(b[j])
#                     j+=1
#             while i<n:
#                 ans.append(a[i])
#                 i+=1
#             while j<m:
#                 ans.append(b[j])
#                 j+=1
#             return ans
        
#         def divide(l):
#             if len(l)<=1:
#                 return l
#             first, last = divide(l[len(l)//2:]), divide(l[:len(l)//2])
            
#             return merge(first, last)
            
#         return divide(nums)

File: 0912-sort-an-array/0912-sort-an-array.java
class Solution {
    public int[] sortArray(int[] nums) {
        return sort(0, nums.length-1, nums);
    }
    
    int[] sort(int i, int j, int[] nums){
        if(i==j) return new int[]{nums[i]};
        int mid = i + (j-i)/2;
        int[] left = sort(i, mid, nums);
        int[] right = sort(mid+1, j, nums);
        
        return merge(left, right, nums);
    }
    
    int[] merge(int[] left, int[] right, int[] nums){
        int l = left.length;
        int r = right.length;
        int[] ans = new int[l+r];
        int i = 0;
        int j = 0;
        int ptr = 0;
        
        
        while(i<l && j<r){
            if(left[i] > right[j])
                ans[ptr] = right[j++];
            else
                ans[ptr] = left[i++];
            ++ptr;
        }
        while(i<l){
            ans[ptr] = left[i++];
            ++ptr;
        }
        while(j<r){
            ans[ptr] = right[j++];
            ++ptr;
        }
        
        return ans;
    }
}